lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -882,158 +877,171 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
-            -- NOTE: It is important that coin selection has the correct
-            -- notion of fees, because it will be used to tell how much
-            -- collateral is needed.
-            , collateralRequirement = colReq
-            , outputsToCover = outs
-            , utxoAvailableForCollateral =
-                UTxOSelection.availableMap utxoSelection
-            , utxoAvailableForInputs = utxoSelection
-            , selectionStrategy = selectionStrategy
-            }
+    mkLedgerTxOut
+        :: RecentEra era
+        -> W.Address
+        -> TokenBundle
+        -> TxOut (ShelleyLedgerEra era)
+    mkLedgerTxOut txOutEra address bundle =
+        case txOutEra of
+            RecentEraBabbage -> W.toBabbageTxOut txOut
+            RecentEraConway -> W.toConwayTxOut txOut
+          where
+            txOut = W.TxOut address bundle
+
+    txPlutusScriptExecutionCost = W.toWallet @W.Coin $
+        if null redeemers
+            then mempty
+            else maxScriptExecutionCost era pp
+
+    collateralRequirement =
+        if txPlutusScriptExecutionCost > W.Coin 0
+            then SelectionCollateralRequired
+            else SelectionCollateralNotRequired
+
+    (positiveBundle, negativeBundle) = posAndNegFromCardanoValue balance
+    TokenBundle positiveAda positiveTokens = positiveBundle
+    TokenBundle negativeAda negativeTokens = negativeBundle
+    adaInOutputs = F.foldMap (TokenBundle.getCoin . view #tokens) outs
+    tokensInOutputs = F.foldMap (TokenBundle.tokens . view #tokens) outs
+    TokenBundle adaInInputs tokensInInputs =
+        UTxOSelection.selectedBalance utxoSelection
+
+    feePerByte = getFeePerByte era pp
+
+    boringFee =
+        calculateMinimumFee
+            (Cardano.AnyCardanoEra (fromRecentEra era))
+            feePerByte
+            txWitnessTag
+            defaultTransactionCtx
+            SelectionSkeleton
+                { skeletonInputCount = UTxOSelection.selectedSize utxoSelection
+                , skeletonOutputs = outs
+                , skeletonChange = []
+                }
+
+    feePadding =
+        let
+            -- Could be made smarter by only padding for the script
+            -- integrity hash when we intend to add one. [ADP-2621]
+            scriptIntegrityHashBytes = 32 + 2
+
+            -- Add padding to allow the fee value to increase.
+            -- Out of caution, assume it can increase by the theoretical
+            -- maximum of 8 bytes ('maximumCostOfIncreasingCoin').
+            --
+            -- NOTE: It's not convenient to import the constant at the
+            -- moment because of the package split.
+            --
+            -- Any overestimation will be reduced by 'distributeSurplus'
+            -- in the final stage of 'balanceTransaction'.
+            extraBytes = 8
+        in
+            W.toWallet . feeOfBytes feePerByte $
+                    extraBytes + scriptIntegrityHashBytes
```

Comment:
This is just how I would write it (so feel free to take with a pinch of salt), but you could save two horizontal lines and one indent by using `where` instead of `let`:

```suggestion
    feePadding 
        = W.toWallet . feeOfBytes feePerByte
        $ extraBytes + scriptIntegrityHashBytes
      where
        -- Could be made smarter by only padding for the script
        -- integrity hash when we intend to add one. [ADP-2621]
        scriptIntegrityHashBytes = 32 + 2

        -- Add padding to allow the fee value to increase.
        -- Out of caution, assume it can increase by the theoretical
        -- maximum of 8 bytes ('maximumCostOfIncreasingCoin').
        --
        -- NOTE: It's not convenient to import the constant at the
        -- moment because of the package split.
        --
        -- Any overestimation will be reduced by 'distributeSurplus'
        -- in the final stage of 'balanceTransaction'.
        extraBytes = 8
```